File: work/build_demo.m
#import <AVFoundation/AVFoundation.h>
#import <CoreGraphics/CoreGraphics.h>
#import <CoreVideo/CoreVideo.h>
#import <Foundation/Foundation.h>
#import <ImageIO/ImageIO.h>

static void Die(NSString *message) {
    fprintf(stderr, "%s\n", message.UTF8String);
    exit(1);
}

static CGImageRef LoadImage(NSString *path) {
    NSURL *url = [NSURL fileURLWithPath:path];
    CGImageSourceRef source = CGImageSourceCreateWithURL((__bridge CFURLRef)url, NULL);
    if (!source) Die([NSString stringWithFormat:@"Cannot load %@", path]);
    CGImageRef image = CGImageSourceCreateImageAtIndex(source, 0, NULL);
    CFRelease(source);
    if (!image) Die([NSString stringWithFormat:@"Cannot decode %@", path]);
    return image;
}

static CVPixelBufferRef MakeFrame(CGImageRef image) {
    NSDictionary *attrs = @{
        (__bridge NSString *)kCVPixelBufferCGImageCompatibilityKey: @YES,
        (__bridge NSString *)kCVPixelBufferCGBitmapContextCompatibilityKey: @YES,
    };
    CVPixelBufferRef buffer = NULL;
    CVReturn status = CVPixelBufferCreate(kCFAllocatorDefault, 1920, 1080,
        kCVPixelFormatType_32ARGB, (__bridge CFDictionaryRef)attrs, &buffer);
    if (status != kCVReturnSuccess || !buffer) Die(@"Cannot allocate video frame");
    CVPixelBufferLockBaseAddress(buffer, 0);
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGContextRef context = CGBitmapContextCreate(CVPixelBufferGetBaseAddress(buffer),
        1920, 1080, 8, CVPixelBufferGetBytesPerRow(buffer),
        colorSpace, (CGBitmapInfo)kCGImageAlphaNoneSkipFirst);
    CGColorSpaceRelease(colorSpace);
    if (!context) Die(@"Cannot create frame context");
    CGContextTranslateCTM(context, 0, 1080);
    CGContextScaleCTM(context, 1, -1);
    CGContextSetInterpolationQuality(context, kCGInterpolationHigh);
    CGContextDrawImage(context, CGRectMake(0, 0, 1920, 1080), image);
    CGContextRelease(context);
    CVPixelBufferUnlockBaseAddress(buffer, 0);
    return buffer;
}

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        if (argc < 4) Die(@"Usage: build_demo output.mp4 seconds-per-slide slide.png [...]");
        NSString *outputPath = [NSString stringWithUTF8String:argv[1]];
        double seconds = atof(argv[2]);
        if (seconds <= 0) Die(@"Slide duration must be positive");
        [[NSFileManager defaultManager] removeItemAtPath:outputPath error:nil];

        NSError *error = nil;
        AVAssetWriter *writer = [[AVAssetWriter alloc]
            initWithURL:[NSURL fileURLWithPath:outputPath]
            fileType:AVFileTypeQuickTimeMovie error:&error];
        if (!writer) Die(error.localizedDescription);
        NSDictionary *settings = @{
            AVVideoCodecKey: AVVideoCodecTypeJPEG,
            AVVideoWidthKey: @1920,
            AVVideoHeightKey: @1080,
            AVVideoCompressionPropertiesKey: @{ AVVideoQualityKey: @0.88 },
        };
        AVAssetWriterInput *input = [AVAssetWriterInput assetWriterInputWithMediaType:AVMediaTypeVideo outputSettings:settings];
        input.expectsMediaDataInRealTime = NO;
        AVAssetWriterInputPixelBufferAdaptor *adaptor =
            [AVAssetWriterInputPixelBufferAdaptor assetWriterInputPixelBufferAdaptorWithAssetWriterInput:input
                sourcePixelBufferAttributes:@{
                    (__bridge NSString *)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32ARGB),
                    (__bridge NSString *)kCVPixelBufferWidthKey: @1920,
                    (__bridge NSString *)kCVPixelBufferHeightKey: @1080,
                }];
        if (![writer canAddInput:input]) Die(@"Cannot add video input");
        [writer addInput:input];
        if (![writer startWriting]) Die(writer.error.localizedDescription);
        [writer startSessionAtSourceTime:kCMTimeZero];

        int64_t frame = 0;
        int framesPerSlide = (int)llround(seconds * 2.0);
        for (int i = 3; i < argc; i++) {
            CGImageRef image = LoadImage([NSString stringWithUTF8String:argv[i]]);
            CVPixelBufferRef pixel = MakeFrame(image);
            CGImageRelease(image);
            for (int j = 0; j < framesPerSlide; j++, frame++) {
                while (!input.readyForMoreMediaData) [NSThread sleepForTimeInterval:0.01];
                if (![adaptor appendPixelBuffer:pixel withPresentationTime:CMTimeMake(frame, 2)]) {
                    Die(writer.error.localizedDescription ?: @"Cannot append frame");
                }
            }
            CVPixelBufferRelease(pixel);
        }
        [input markAsFinished];
        dispatch_semaphore_t done = dispatch_semaphore_create(0);
        [writer finishWritingWithCompletionHandler:^{ dispatch_semaphore_signal(done); }];
        dispatch_semaphore_wait(done, DISPATCH_TIME_FOREVER);
        if (writer.status != AVAssetWriterStatusCompleted) Die(writer.error.localizedDescription ?: @"Video writing failed");
        printf("%s\n", outputPath.UTF8String);
    }
    return 0;
}

File: work/build_demo.swift
import AVFoundation
import CoreGraphics
import Foundation
import ImageIO

struct Segment {
    let imageURL: URL
    let audioURL: URL
    let duration: CMTime
}

func fail(_ message: String) -> Never {
    FileHandle.standardError.write(Data((message + "\n").utf8))
    exit(1)
}

func loadImage(_ url: URL) -> CGImage {
    guard let source = CGImageSourceCreateWithURL(url as CFURL, nil),
          let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
        fail("Cannot load slide image: \(url.path)")
    }
    return image
}

func audioDuration(_ url: URL) -> CMTime {
    let asset = AVURLAsset(url: url)
    let seconds = asset.duration.seconds
    guard seconds.isFinite && seconds > 0 else { fail("Invalid audio duration: \(url.path)") }
    return CMTime(seconds: seconds + 0.35, preferredTimescale: 600)
}

func makePixelBuffer(_ image: CGImage, width: Int, height: Int) -> CVPixelBuffer {
    var buffer: CVPixelBuffer?
    let attrs: [CFString: Any] = [
        kCVPixelBufferCGImageCompatibilityKey: true,
        kCVPixelBufferCGBitmapContextCompatibilityKey: true,
    ]
    let status = CVPixelBufferCreate(kCFAllocatorDefault, width, height, kCVPixelFormatType_32ARGB, attrs as CFDictionary, &buffer)
    guard status == kCVReturnSuccess, let pixelBuffer = buffer else { fail("Cannot create pixel buffer") }
    CVPixelBufferLockBaseAddress(pixelBuffer, [])
    defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, []) }
    guard let context = CGContext(
        data: CVPixelBufferGetBaseAddress(pixelBuffer), width: width, height: height,
        bitsPerComponent: 8, bytesPerRow: CVPixelBufferGetBytesPerRow(pixelBuffer),
        space: CGColorSpaceCreateDeviceRGB(),
        bitmapInfo: CGImageAlphaInfo.noneSkipFirst.rawValue
    ) else { fail("Cannot create drawing context") }
    context.setFillColor(CGColor(gray: 0.96, alpha: 1))
    context.fill(CGRect(x: 0, y: 0, width: width, height: height))
    context.translateBy(x: 0, y: CGFloat(height))
    context.scaleBy(x: 1, y: -1)
    context.interpolationQuality = .high
    context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
    return pixelBuffer
}

func buildVideo(segments: [Segment], output: URL) throws -> [CMTime] {
    try? FileManager.default.removeItem(at: output)
    let writer = try AVAssetWriter(outputURL: output, fileType: .mov)
    let settings: [String: Any] = [
        AVVideoCodecKey: AVVideoCodecType.h264,
        AVVideoWidthKey: 1920,
        AVVideoHeightKey: 1080,
        AVVideoCompressionPropertiesKey: [AVVideoAverageBitRateKey: 5_000_000],
    ]
    let input = AVAssetWriterInput(mediaType: .video, outputSettings: settings)
    input.expectsMediaDataInRealTime = false
    let adaptor = AVAssetWriterInputPixelBufferAdaptor(
        assetWriterInput: input,
        sourcePixelBufferAttributes: [
            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32ARGB,
            kCVPixelBufferWidthKey as String: 1920,
            kCVPixelBufferHeightKey as String: 1080,
        ]
    )
    guard writer.canAdd(input) else { fail("Cannot add video input") }
    writer.add(input)
    guard writer.startWriting() else { fail("Cannot start video writer: \(writer.error?.localizedDescription ?? "unknown")") }
    writer.startSession(atSourceTime: .zero)
    var starts: [CMTime] = []
    var cursor = CMTime.zero
    let frameStep = CMTime(value: 1, timescale: 2)
    for segment in segments {
        starts.append(cursor)
        let pixel = makePixelBuffer(loadImage(segment.imageURL), width: 1920, height: 1080)
        let end = cursor + segment.duration
        var frameTime = cursor
        while frameTime < end {
            while !input.isReadyForMoreMediaData { Thread.sleep(forTimeInterval: 0.01) }
            guard adaptor.append(pixel, withPresentationTime: frameTime) else {
                fail("Cannot append frame: \(writer.error?.localizedDescription ?? "unknown")")
            }
            frameTime = frameTime + frameStep
        }
        cursor = end
    }
    input.markAsFinished()
    let semaphore = DispatchSemaphore(value: 0)
    writer.finishWriting { semaphore.signal() }
    semaphore.wait()
    guard writer.status == .completed else { fail("Video writing failed: \(writer.error?.localizedDescription ?? "unknown")") }
    return starts
}

func combine(videoURL: URL, segments: [Segment], starts: [CMTime], output: URL) throws {
    try? FileManager.default.removeItem(at: output)
    let composition = AVMutableComposition()
    let videoAsset = AVURLAsset(url: videoURL)
    guard let sourceVideo = videoAsset.tracks(withMediaType: .video).first,
          let videoTrack = composition.addMutableTrack(withMediaType: .video, preferredTrackID: kCMPersistentTrackID_Invalid) else {
        fail("Cannot read generated video")
    }
    try videoTrack.insertTimeRange(CMTimeRange(start: .zero, duration: videoAsset.duration), of: sourceVideo, at: .zero)
    for (index, segment) in segments.enumerated() {
        let audioAsset = AVURLAsset(url: segment.audioURL)
        guard let sourceAudio = audioAsset.tracks(withMediaType: .audio).first,
              let audioTrack = composition.addMutableTrack(withMediaType: .audio, preferredTrackID: kCMPersistentTrackID_Invalid) else {
            fail("Cannot read narration: \(segment.audioURL.path)")
        }
        let usable = min(audioAsset.duration, segment.duration)
        try audioTrack.insertTimeRange(CMTimeRange(start: .zero, duration: usable), of: sourceAudio, at: starts[index])
    }
    guard let export = AVAssetExportSession(asset: composition, presetName: AVAssetExportPresetHighestQuality) else {
        fail("Cannot create export session")
    }
    export.outputURL = output
    export.outputFileType = .mp4
    export.shouldOptimizeForNetworkUse = true
    let semaphore = DispatchSemaphore(value: 0)
    export.exportAsynchronously { semaphore.signal() }
    semaphore.wait()
    guard export.status == .completed else { fail("Export failed: \(export.error?.localizedDescription ?? "unknown")") }
}

let args = CommandLine.arguments
guard args.count >= 5, (args.count - 3) % 2 == 0 else {
    fail("Usage: build_demo temp.mov output.mp4 slide.png narration.aiff [slide.png narration.aiff ...]")
}
let temporary = URL(fileURLWithPath: args[1])
let output = URL(fileURLWithPath: args[2])
var segments: [Segment] = []
var index = 3
while index < args.count {
    let image = URL(fileURLWithPath: args[index])
    let audio = URL(fileURLWithPath: args[index + 1])
    segments.append(Segment(imageURL: image, audioURL: audio, duration: audioDuration(audio)))
    index += 2
}
let starts = try buildVideo(segments: segments, output: temporary)
try combine(videoURL: temporary, segments: segments, starts: starts, output: output)
try? FileManager.default.removeItem(at: temporary)
print(output.path)
